include/seqan3/contrib/parallel/reader_writer_latch.hpp

```diff
@@ -0,0 +1,323 @@
+        assert(wcount.get() >= 0);
+    }
+
+    /*!\brief Constructs the reader_writer_latch with the writer count and the associated data structure.
+     * \tparam concurrent_t The type of the concurrent data structure. Must have a member function `close`.
+     * \param wcount The expected number of producer threads.
+     * \param ds     The concurrent data structure.
+     *
+     * \details
+     *
+     * Delegates to reader_writer_latch::reader_writer_latch(reader_count, writer_count, concurrent_t) with reader
+     * count set to 0.
+     */
+    template <typename concurrent_t>
+        requires requires { std::declval<concurrent_t>().close(); }  // requires closable queue.
+    reader_writer_latch(writer_count const wcount, concurrent_t & queue) :
+        reader_writer_latch{wcount, reader_count{0}, queue}
+    {}
+
+    /*!\brief Constructs the reader_writer_latch with the reader count and the associated data structure.
+     * \tparam concurrent_t The type of the concurrent data structure. Must have a member function `close`.
+     * \param rcount The expected number of consumer threads.
+     * \param ds     The concurrent data structure.
+     *
+     * \details
+     *
+     * Delegates to reader_writer_latch::reader_writer_latch(reader_count, writer_count, concurrent_t) with writer
+     * count set to 0.
+     */
+    template <typename concurrent_t>
+        requires requires { std::declval<concurrent_t>().close(); }  // requires closable queue.
+    reader_writer_latch(reader_count const rcount, concurrent_t & queue) :
+        reader_writer_latch{writer_count{0}, rcount, queue}
+    {}
+    //!\}
+
+    /*!\brief Atomically decrements writer counter by one and blocks the calling thread.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for producer threads and blocks the calling thread until all participating
+     * threads have arrived. If all participating producer threads arrived at the synchronisation point, one of the
+     * threads will invoke the completion function.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive_and_wait() noexcept
+    {
+        writer_latch.arrive_and_wait();
+
+        std::call_once(flag, completion_fn);
+    }
+
+    /*!\brief Atomically decrements writer counter by one.
+     *
+     * \details
+     *
+     * Arrives at the synchronisation point for producer threads without blocking the calling thread.
+     * If all participating producer threads arrived at the synchronisation point, one of the
+     * threads will invoke the completion function.
+     *
+     * ### Exception
+     *
+     * no-throw
+     *
+     * ### Thread safety
+     *
+     * thread-safe
+     */
+    void writer_arrive() noexcept
+    {
+        writer_latch.arrive();
+
+        if (writer_latch.try_wait())
+            std::call_once(flag, completion_fn);
```

Comment:
what happens if one thread invokes `writer_arrive_and_wait` and the other `writer_arrive`.

Is it possible that one thread executes `std::call_once(flag, completion_fn);` within `writer_arrive` and another thread did not execute `writer_arrive_and_wait`, yet?

`completion_fn` aka `ds.close` blocking? That means `writer_arrive` may be a `wait` function and has potential for death locks?

WARNING: This may be a stupid question.

Reply:
Good question: Assume t1 enters `writer_arrive_and_wait` while t2 enters at the same time`writer_arrive`. Both are participating threads so the latch counter must be at least 2. If t1 has not executed yet t2 will decrement the counter and try_wait returns false. If t1 has executed in the meantime (until t2 asks try_wait) t2 will call the completion, but t1 is already finished. If t1 has executed but t2 not, then t1 will wait until t2 executes. If t2 executes then again either t2 or t1 will call the termination. The call_once is thread_safe and guarantees that only one thread is getting the privilege to call the completion phase. I think this is fairly safe.